### PR TITLE
Simplification of property groups usage

### PR DESCRIPTION
**Pull request by neilkakkar**
## Changes

`property_groups` are the new properties and every place should **replace** `properties`, not condition on them.

~~Experimenting right now, this is just a PoC.~~

Have split it into manageable chunks. This sets up step 1 of migration plan here: https://github.com/PostHog/posthog/issues/8670, subsequent PRs will tackle:

1. simplifying filters
2. column optimizer and person property push down
3. entity properties
4. ???

## How did you test this code?

Existing unit tests

### CODE REVIEW

File: posthog/models/filters/mixins/property.py [hunk 1]
```diff
@@ -22,14 +22,17 @@ def properties(self) -> List[Property]:
...
+            # old style dict properties or a list of properties
+            return self._parse_properties(loaded_props)
 
     @cached_property
     def property_groups(self) -> PropertyGroup:
-        _props = self._data.get(PROPERTY_GROUPS, None)
-
-        if not _props:
-            return PropertyGroup(type=PropertyOperatorType.AND, groups=self.properties)
+        _props = self._data.get(PROPERTIES)
```
Comment (neilkakkar) [inline posthog/models/filters/mixins/property.py:35]: Keeping the API the same^^

and at the same time, parsing such that `property_groups` are the new properties. Eventually, we should remove all occurences of properties, to be replaced with property_groups.

File: posthog/models/property.py [hunk 2]
```diff
@@ -152,6 +152,26 @@ def __init__(self, type: PropertyOperatorType, groups: Union[List[Property], Lis
         self.type = type
         self.groups = groups
 
+    def combine_property_groups(
```
Comment (neilkakkar) [inline posthog/models/property.py:155]: TODO: tests!

**Comment by neilkakkar** [comment]
Think this is in a decent place. Here's the new deal: `filter.properties` should go away completely in app. It hasn't yet, because step-wise migration. (when they're gone, can remove the property from filter mixin)

`PropertyGroup` is the big object which would, in the future, have: behavioural properties, regular properties, and grouped AND/OR properties.

To get here, I've removed filter.properties in _some_ places, while keeping PR size manageable.

Also, as @macobo mentioned, not extending API, just using the "properties" key for new style properties as well.

@macobo : Let me know if this approach makes sense / isn't what you had in mind. I didn't go the `AnyProperty = Union[PropertyGroup | Property]` route, because `PropertyGroup` is kinda this grand property. Using property_group everywhere helps reach the same typing goals, as things that don't accept property groups will scream, which we should fix.

File: posthog/models/property.py (hunk 2 above)
Comment (EDsCODE) [reply posthog/models/property.py:155]: Where is this used?

**Review by EDsCODE — APPROVED** [review]

File: posthog/models/property.py (hunk 2 above)
Comment (neilkakkar) [reply posthog/models/property.py:155]: Huh, this one isn't!

Tests just for `combine_properties` then, I guess. 😅

**Comment by EDsCODE** [comment]
Was working through the simplification logic a bit more and there's a few more areas that could use cleanup. Some trivial merge conflicts I had to introduce for consistency on the base PR. 

`ee/clickhouse/queries/breakdown_props.py` line 55  needs to use filter.property_groups too. Worth double checking if there's any querying elsewhere that needs this updated



**RE: simplify logic** 
I made the changes you suggested "return property_groups and empty properties". However, it caused tons of test issues unless the rest of the logic for query building also defaults to property_groups which is handled in this PR! So, I'd recommend going to `posthog/models/filters/mixin/simplify.py` and removing:

- `"properties": result.properties + team.test_account_filters,` from line 31
- `properties = self._simplify_properties(team, result.properties, **kwargs)  # type: ignore` line 41
- `properties.append(self._group_set_property(cast(int, result.aggregation_group_type_index)))  # type: ignore` line 44
- and changing properties to return empty list

Sorry for the janky suggestions. Not sure how to suggest edits outside the scope of what's touched in the PR. Those edits to simplify will return simplified property_groups only and if all the if statements are properly removed in this PR, there should be minimal test errors 🙏

**Comment by macobo** [comment]
It's really hard to see whether this covers everything and is correct since some code is using `.property_groups`, others `.properties` hence making this change tricky. I think this can be avoided with some clever refactoring - let's discuss if you're interested?

File: posthog/models/filters/mixins/property.py [hunk 3]
```diff
@@ -39,7 +42,16 @@ def property_groups(self) -> PropertyGroup:
         else:
             loaded_props = _props
 
-        return self._parse_property_group(loaded_props)
+        # if grouped properties
+        if isinstance(loaded_props, dict) and "type" in loaded_props and "groups" in loaded_props:
+            try:
+                return self._parse_property_group(loaded_props)
+            except Exception as e:
+                # TODO: make exception class precise?
```
Comment (macobo) [inline posthog/models/filters/mixins/property.py:50]: This is IMO a requirement for merging this. - we should not be catching generic exception ever.

File: posthog/models/filters/mixins/property.py [hunk 4]
```diff
@@ -22,14 +22,17 @@ def properties(self) -> List[Property]:
         else:
             loaded_props = _props
 
-        return self._parse_properties(loaded_props)
+        # if grouped properties
+        if isinstance(loaded_props, dict) and "type" in loaded_props and "groups" in loaded_props:
```
Comment (macobo) [inline posthog/models/filters/mixins/property.py:26]: We have two divergent paths here. IMO the saner way is to change `.properties` and not introduce `.property_groups` at all (we can always rename later if needed) but if we _really_ need both one should directly depend on the other.

File: posthog/models/property.py (hunk 2 above)
Comment (macobo) [reply posthog/models/property.py:155]: What does combine even mean, why do we need two different methods for properties vs property groups?

**Comment by neilkakkar** [comment]
@EDsCODE this is getting too big, so [I made a migration plan outlining our strategy for making this happen.](https://github.com/PostHog/posthog/issues/8670)

Will merge this PR into yours, and do the proper review there, otherwise getting hard to handle what's changing and what isn't. Will possibly make the change smaller, if it means touching fewer things that can be tested well in isolation.

**Comment by neilkakkar** [comment]
(actually, since this is based on top of yours, and yours can't go in until this is merged into yours, I'll just close yours and change the base here to master?)

File: ee/clickhouse/models/test/test_filters.py [hunk 5]
```diff
@@ -62,32 +62,56 @@ def test_simplify_cohorts(self):
         filter = Filter(data={"properties": [{"type": "cohort", "key": "id", "value": cohort.pk}]})
 
         self.assertEqual(
-            filter.simplify(self.team).properties_to_dict(),
-            {"properties": [{"key": "email", "operator": "icontains", "value": ".com", "type": "person"},]},
+            filter.simplify(self.team).properties_to_dict(), {},
```
Comment (neilkakkar) [inline ee/clickhouse/models/test/test_filters.py:65]: think I'll carve simplify changes out to a separate PR. We want these tests for both old and new style properties, for the time being. And until migration completes, simplify using old properties is perfectly fine.

**Comment by neilkakkar** [comment]
Have made this into a small chunk, which sets up the migration plan, and upgrades all event queries.

Ready for a review, @macobo , @EDsCODE .

File: ee/clickhouse/models/property.py [hunk 6]
```diff
@@ -158,9 +161,17 @@ def parse_prop_clauses(
                     f"{property_operator} 0 = 13"
                 )  # If cohort doesn't exist, nothing can match, unless an OR operator is used
             else:
-                person_id_query, cohort_filter_params = format_filter_query(cohort, idx)
-                params = {**params, **cohort_filter_params}
-                final.append(f"{property_operator} {table_name}distinct_id IN ({person_id_query})")
+
+                if person_properties_mode == PersonPropertiesMode.EXCLUDE:
```
Comment (macobo) [inline ee/clickhouse/models/property.py:165]: Q: Context on why cohorts needed to change this way?

File: ee/clickhouse/queries/event_query.py [hunk 7]
```diff
@@ -95,7 +100,7 @@ def _determine_should_join_persons(self) -> None:
 
         # :KLUDGE: The following is mostly making sure if cohorts are included as well.
         #   Can be simplified significantly after https://github.com/PostHog/posthog/issues/5854
-        if any(self._should_property_join_persons(prop) for prop in self._filter.properties):
+        if any(self._should_property_join_persons(prop) for prop in self._filter.property_groups_flat):
```
Comment (macobo) [inline ee/clickhouse/queries/event_query.py:103]: Is property_groups_flat a purely transitional helper or something we'll keep around?

File: ee/clickhouse/queries/trends/breakdown.py [hunk 8]
```diff
@@ -95,7 +97,7 @@ def get_query(self) -> Tuple[str, Dict, Callable]:
             "parsed_date_to": parsed_date_to,
             "actions_query": "AND {}".format(action_query) if action_query else "",
             "event_filter": "AND event = %(event)s" if not action_query else "",
-            "filters": prop_filters if props_to_filter else "",
+            "filters": prop_filters if props_to_filter.groups else "",
```
Comment (macobo) [inline ee/clickhouse/queries/trends/breakdown.py:100]: Hmm. Took me a second to grasp what this value `.groups` is. Perhaps we shouldn't introduce that verb at all here (since it causes confusion with e.g. group analytics) and instead name it `values` or `property_groups`?

File: ee/clickhouse/models/property.py (hunk 6 above)
Comment (EDsCODE) [reply ee/clickhouse/models/property.py:165]: ClickhouseEventQuery was parsing props in a special way in `_get_props` and used an extra util function `_get_cohort_subquery` which was doing almost the same thing as parse_prop_clause handling of cohort filters. This wouldn't work since we're now relying on `parse_prop_grouped_clauses` for the entire filter parsing

File: ee/clickhouse/queries/trends/breakdown.py (hunk 8 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/trends/breakdown.py:100]: agreed!

File: posthog/models/filters/mixins/property.py (hunk 4 above)
Comment (macobo) [reply posthog/models/filters/mixins/property.py:26]: Ping/thoughts on this? I'd implement this in terms of `.property_groups` based on our call.

File: ee/clickhouse/queries/event_query.py (hunk 7 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/event_query.py:103]: It should stick around because it's necessary for digesting what properties are being used so that any functions that are checking on properties being filtered on don't need to keep redoing a tree traversal

File: ee/clickhouse/queries/trends/breakdown.py (hunk 8 above)
Comment (neilkakkar) [reply ee/clickhouse/queries/trends/breakdown.py:100]: Yeahh, definitely.

File: posthog/models/filters/mixins/property.py (hunk 4 above)
Comment (EDsCODE) [reply posthog/models/filters/mixins/property.py:26]: Sorry, what's being asked here? I thought the plan now is to transition properties -> property_groups so it is a one way dependence. Also, every other property filtering area should remain the same until we make those changes. We're not changing entity properties to use or filtering yet so it can remain the same with the list of properties as long as we combine them with the property_group correctly during query building

**Review by EDsCODE — COMMENTED** [review]
I think the migration plan makes sense but I'm a bit puzzled on the ordering

Is this PR meant to get the property_groups into a working state so that the frontend can be test? Or are we now supposed to proceed and finish converting all related areas that still use filter.properties into filter.property_groups while pointing those branches into this branch?

Because, for something like `get_breakdown_prop_values`, I don't think it'll work when you pass in a property_group since it directly relies on filter.properties still and filter.properties will return empty if property_groups are parsed

**Comment by neilkakkar** [comment]
Good question!

New-format properties will not work for all cases until the migration is complete -> this is true. Can make it a bit annoying to test all cases in the frontend, until migration is complete.

This PR will go in on approval, because it allows old properties to keep working fine.

The goal of the migration (in my eyes) is to make the transition to new properties smooth: where we tackle each chunk on its own. For example, with breakdown_props, when we update to using `property_groups`, ideally we'll have a few test cases with the new `property_groups` set in `breakdown_cases.py`. 

Personally, I felt like there were too many things to keep track of in my head if I tried doing them all in one PR. (For example, I did set breakdown_prop properties to the new ones, but it was such a small change in an otherwise huge PR that I was averse to adding tests for it and making the PR even huger, and since it was already huge, it was easy to miss during reviews that there aren't new tests for this bit).

But now, this PR will go in, and then every chunk can go in independently, with its own good tests. There's atleast 4 such chunks already.

**Review by EDsCODE — COMMENTED** [review]
The init of `Filter` is going to cause trouble. It will never parse a properties payload that uses groups

**Comment by neilkakkar** [comment]
Why do you think so? 🤔

**Comment by EDsCODE** [comment]
I don't think so I was misreading the validation T_T

File: ee/clickhouse/queries/trends/breakdown.py (hunk 8 above)
Comment (neilkakkar) [reply ee/clickhouse/queries/trends/breakdown.py:100]: (will do in a follow up, makes other things hard to parse in here)
